File: services/ims-api/internal/models/messaging.go
package models

import (
	"encoding/json"
	"time"
)

// ThreadType represents the type of message thread
type ThreadType string

const (
	ThreadTypeGeneral  ThreadType = "general"
	ThreadTypeIncident ThreadType = "incident"
	ThreadTypeLivechat ThreadType = "livechat"
)

// ThreadStatus represents the status of a message thread
type ThreadStatus string

const (
	ThreadStatusOpen     ThreadStatus = "open"
	ThreadStatusClosed   ThreadStatus = "closed"
	ThreadStatusArchived ThreadStatus = "archived"
)

// ContentType represents the type of message content
type ContentType string

const (
	ContentTypeText       ContentType = "text"
	ContentTypeSystem     ContentType = "system"
	ContentTypeAttachment ContentType = "attachment"
)

// MessageThread represents a conversation thread
type MessageThread struct {
	ID                 string       `json:"id"`
	TenantID           string       `json:"tenantId"`
	SchoolID           string       `json:"schoolId"`
	Subject            string       `json:"subject"`
	ThreadType         ThreadType   `json:"threadType"`
	Status             ThreadStatus `json:"status"`
	IncidentID         *string      `json:"incidentId,omitempty"`
	CreatedBy          string       `json:"createdBy"`
	CreatedByRole      string       `json:"createdByRole"`
	CreatedByName      string       `json:"createdByName"`
	MessageCount       int          `json:"messageCount"`
	UnreadCountSchool  int          `json:"unreadCountSchool"`
	UnreadCountSupport int          `json:"unreadCountSupport"`
	LastMessageAt      *time.Time   `json:"lastMessageAt,omitempty"`
	CreatedAt          time.Time    `json:"createdAt"`
	UpdatedAt          time.Time    `json:"updatedAt"`
	ClosedAt           *time.Time   `json:"closedAt,omitempty"`

	// Denormalized fields for display
	SchoolName  string   `json:"schoolName,omitempty"`
	LastMessage *Message `json:"lastMessage,omitempty"`
}

// Message represents an individual message in a thread
type Message struct {
	ID          string            `json:"id"`
	TenantID    string            `json:"tenantId"`
	ThreadID    string            `json:"threadId"`
	SenderID    string            `json:"senderId"`
	SenderName  string            `json:"senderName"`
	SenderRole  string            `json:"senderRole"`
	Content     string            `json:"content"`
	ContentType ContentType       `json:"contentType"`
	Metadata    map[string]any    `json:"metadata,omitempty"`
	EditedAt    *time.Time        `json:"editedAt,omitempty"`
	DeletedAt   *time.Time        `json:"deletedAt,omitempty"`
	CreatedAt   time.Time         `json:"createdAt"`
	Attachments []MessageAttachment `json:"attachments,omitempty"`
}

// MessageAttachment represents a file attached to a message
type MessageAttachment struct {
	ID           string    `json:"id"`
	TenantID     string    `json:"tenantId"`
	MessageID    string    `json:"messageId"`
	FileName     string    `json:"fileName"`
	ContentType  string    `json:"contentType"`
	SizeBytes    int64     `json:"sizeBytes"`
	ObjectKey    string    `json:"objectKey"`
	ThumbnailKey *string   `json:"thumbnailKey,omitempty"`
	CreatedAt    time.Time `json:"createdAt"`

	// Presigned URLs (populated on read)
	DownloadURL  string `json:"downloadUrl,omitempty"`
	ThumbnailURL string `json:"thumbnailUrl,omitempty"`
}

// ThreadParticipant represents a user participating in a thread
type ThreadParticipant struct {
	ThreadID string     `json:"threadId"`
	UserID   string     `json:"userId"`
	UserName string     `json:"userName"`
	UserRole string     `json:"userRole"`
	JoinedAt time.Time  `json:"joinedAt"`
	LeftAt   *time.Time `json:"leftAt,omitempty"`
}

// ReadReceipt tracks when a user last read a thread
type ReadReceipt struct {
	ThreadID          string    `json:"threadId"`
	UserID            string    `json:"userId"`
	LastReadMessageID string    `json:"lastReadMessageId"`
	LastReadAt        time.Time `json:"lastReadAt"`
}

// UnreadCounts holds unread message counts for a user
type UnreadCounts struct {
	Threads  int `json:"threads"`
	Messages int `json:"messages"`
}

// CreateThreadRequest represents the request to create a new thread
type CreateThreadRequest struct {
	Subject        string   `json:"subject"`
	IncidentID     *string  `json:"incidentId,omitempty"`
	InitialMessage string   `json:"initialMessage"`
	Attachments    []string `json:"attachments,omitempty"`
	SchoolID       string   `json:"schoolId,omitempty"` // Optional, for support agents creating threads
}

// CreateMessageRequest represents the request to send a message
type CreateMessageRequest struct {
	Content     string   `json:"content"`
	Attachments []string `json:"attachments,omitempty"`
}

// UpdateThreadStatusRequest represents the request to update thread status
type UpdateThreadStatusRequest struct {
	Status ThreadStatus `json:"status"`
}

// MarkReadRequest represents the request to mark a thread as read
type MarkReadRequest struct {
	LastMessageID string `json:"lastMessageId"`
}

// ThreadsListResponse represents a paginated list of threads
type ThreadsListResponse struct {
	Items      []MessageThread `json:"items"`
	NextCursor string          `json:"nextCursor,omitempty"`
}

// MessagesListResponse represents a paginated list of messages
type MessagesListResponse struct {
	Items      []Message `json:"items"`
	NextCursor string    `json:"nextCursor,omitempty"`
}

// ThreadDetailResponse represents a thread with its messages
type ThreadDetailResponse struct {
	Thread       MessageThread       `json:"thread"`
	Messages     []Message           `json:"messages"`
	Participants []ThreadParticipant `json:"participants,omitempty"`
}

// CreateThreadResponse represents the response when creating a thread
type CreateThreadResponse struct {
	Thread  MessageThread `json:"thread"`
	Message Message       `json:"message"`
}

// CreateMessageResponse represents the response when creating a message
type CreateMessageResponse struct {
	Message Message `json:"message"`
}

// UploadURLRequest represents the request to get an upload URL
type UploadURLRequest struct {
	FileName    string `json:"fileName"`
	ContentType string `json:"contentType"`
	SizeBytes   int64  `json:"sizeBytes"`
}

// UploadURLResponse represents the response with an upload URL
type UploadURLResponse struct {
	UploadURL     string `json:"uploadUrl"`
	AttachmentRef string `json:"attachmentRef"`
}

// MessageSearchResult represents a message in search results
type MessageSearchResult struct {
	Message    Message       `json:"message"`
	Thread     MessageThread `json:"thread"`
	Highlights []string      `json:"highlights,omitempty"`
}

// SearchResponse represents search results
type SearchResponse struct {
	Items []MessageSearchResult `json:"items"`
	Total int                   `json:"total"`
}

// MessagingAnalytics represents messaging analytics data
type MessagingAnalytics struct {
	ThreadsCreated    int     `json:"threadsCreated"`
	MessagesSent      int     `json:"messagesSent"`
	AvgResponseTime   float64 `json:"avgResponseTimeMinutes"`
	ChatSessions      int     `json:"chatSessions"`
	AvgChatRating     float64 `json:"avgChatRating"`
	ActiveThreads     int     `json:"activeThreads"`
	ClosedThreads     int     `json:"closedThreads"`
	AvgMessagesThread float64 `json:"avgMessagesPerThread"`
}

// MetadataBytes returns metadata as JSON bytes
func (m *Message) MetadataBytes() []byte {
	if m.Metadata == nil {
		return []byte("{}")
	}
	b, _ := json.Marshal(m.Metadata)
	return b
}

// IsSchoolRole returns true if the role is a school contact
func IsSchoolRole(role string) bool {
	return role == "ssp_school_contact"
}

// IsSupportRole returns true if the role is a support/admin role
func IsSupportRole(role string) bool {
	return role == "ssp_admin" || role == "ssp_support_agent"
}

File: dashboard/src/types/messaging.ts
// Thread types
export type ThreadType = 'general' | 'incident' | 'livechat';
export type ThreadStatus = 'open' | 'closed' | 'archived';
export type ContentType = 'text' | 'system' | 'attachment';

// Message Thread
export interface MessageThread {
  id: string;
  tenantId: string;
  schoolId: string;
  subject: string;
  threadType: ThreadType;
  status: ThreadStatus;
  incidentId?: string;
  createdBy: string;
  createdByRole: string;
  createdByName: string;
  messageCount: number;
  unreadCountSchool: number;
  unreadCountSupport: number;
  lastMessageAt?: string;
  createdAt: string;
  updatedAt: string;
  closedAt?: string;
  schoolName?: string;
  lastMessage?: Message;
}

// Message
export interface Message {
  id: string;
  tenantId: string;
  threadId: string;
  senderId: string;
  senderName: string;
  senderRole: string;
  content: string;
  contentType: ContentType;
  metadata?: Record<string, unknown>;
  editedAt?: string;
  deletedAt?: string;
  createdAt: string;
  attachments?: MessageAttachment[];
}

// Message Attachment
export interface MessageAttachment {
  id: string;
  tenantId: string;
  messageId: string;
  fileName: string;
  contentType: string;
  sizeBytes: number;
  objectKey: string;
  thumbnailKey?: string;
  downloadUrl?: string;
  thumbnailUrl?: string;
  createdAt: string;
}

// Thread Participant
export interface ThreadParticipant {
  threadId: string;
  userId: string;
  userName: string;
  userRole: string;
  joinedAt: string;
  leftAt?: string;
}

// Read Receipt
export interface ReadReceipt {
  threadId: string;
  userId: string;
  lastReadMessageId: string;
  lastReadAt: string;
}

// Unread Counts
export interface UnreadCounts {
  threads: number;
  messages: number;
  total: number;
}

// Request types
export interface CreateThreadRequest {
  subject: string;
  incidentId?: string;
  initialMessage: string;
  attachments?: string[];
  schoolId?: string;
}

export interface CreateMessageRequest {
  content: string;
  attachments?: string[];
}

export interface UpdateThreadStatusRequest {
  status: ThreadStatus;
}

export interface MarkReadRequest {
  lastMessageId: string;
}

export interface UploadURLRequest {
  fileName: string;
  contentType: string;
  sizeBytes: number;
}

// Response types
export interface ThreadsListResponse {
  items: MessageThread[];
  nextCursor?: string;
}

export interface MessagesListResponse {
  items: Message[];
  nextCursor?: string;
}

export interface ThreadDetailResponse {
  thread: MessageThread;
  messages: Message[];
  participants?: ThreadParticipant[];
}

export interface CreateThreadResponse {
  thread: MessageThread;
  message: Message;
}

export interface CreateMessageResponse {
  message: Message;
}

export interface UploadURLResponse {
  uploadUrl: string;
  attachmentRef: string;
}

export interface MessageSearchResult {
  message: Message;
  thread: MessageThread;
  highlights?: string[];
}

export interface SearchResponse {
  items: MessageSearchResult[];
  total: number;
}

// Analytics
export interface MessagingAnalytics {
  threadsCreated: number;
  messagesSent: number;
  avgResponseTimeMinutes: number;
  chatSessions: number;
  avgChatRating: number;
  activeThreads: number;
  closedThreads: number;
  avgMessagesPerThread: number;
}

// WebSocket message types for real-time updates
export interface ChatMessageEvent {
  type: 'chat_message';
  payload: {
    threadId: string;
    message: Message;
  };
}

export interface ChatTypingEvent {
  type: 'chat_typing';
  payload: {
    threadId: string;
    userId: string;
    userName: string;
    isTyping: boolean;
  };
}

export interface ChatReadEvent {
  type: 'chat_read';
  payload: {
    threadId: string;
    userId: string;
    lastMessageId: string;
    readAt: string;
  };
}

// List params
export interface ThreadsListParams {
  status?: string;
  incidentId?: string;
  q?: string;
  cursor?: string;
  limit?: number;
}
